tests/function-controller/testsuite/scenarios/gitauth.go

```diff
@@ -0,0 +1,206 @@
+		}
+		steps = append(steps, testSteps)
+	}
+	return step.NewSerialTestRunner(logf, "Test Git function authentication",
+		teststep.NewNamespaceStep("Create test namespace", coreCli, genericContainer),
+		step.NewParallelRunner(logf, "fn_tests",
+			steps...)), nil
+}
+
+func setupSharedContainer(restConfig *rest.Config, cfg testsuite.Config, logf *logrus.Entry) (shared.Container, error) {
+	currentDate := time.Now()
+	cfg.Namespace = fmt.Sprintf("%s-%dh-%dm-%d", "test-serverless-gitauth", currentDate.Hour(), currentDate.Minute(), rand.Int())
+
+	dynamicCli, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return shared.Container{}, errors.Wrapf(err, "while creating dynamic client")
+	}
+
+	return shared.Container{
+		DynamicCli:  dynamicCli,
+		Namespace:   cfg.Namespace,
+		WaitTimeout: cfg.WaitTimeout,
+		Verbose:     cfg.Verbose,
+		Log:         logf,
+	}, nil
+}
+func gitAuthFunctionTestSteps(genericContainer shared.Container, tr testRepo, poll poller.Poller) (step.Step, error) {
+	genericContainer.Log.Infof("Testing Git Function in namespace: %s", genericContainer.Namespace)
+
+	secret := secret.NewSecret(tr.auth.SecretName, genericContainer)
+
+	gitCfg, err := gitops.NewGitopsConfig(
+		tr.name,
+		fmt.Sprintf("%s-repo", tr.name),
+		"",
+		"",
+		genericContainer,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while creating Git config")
+	}
+
+	data, err := tr.authSecretDataFunc(tr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while generating secret data")
+	}
+
+	return step.NewSerialTestRunner(genericContainer.Log, fmt.Sprintf("%s Function auth test", tr.provider),
+		teststep.CreateSecret(
+			genericContainer.Log,
+			secret,
+			fmt.Sprintf("Create %s Auth Secret", tr.provider),
+			data),
+		teststep.NewCreateGitRepository(
+			genericContainer.Log,
+			gitCfg.Repo,
+			fmt.Sprintf("Create GitRepository for %s", tr.provider),
+			gitops.AuthRepositorySpecWithURL(
+				tr.url,
+				tr.auth,
+			)),
+		teststep.CreateFunction(
+			genericContainer.Log,
+			gitCfg.Fn,
+			fmt.Sprintf("Create %s Function", tr.provider),
+			gitops.GitopsFunction(
+				gitCfg.RepoName,
+				tr.baseDir,
+				tr.reference,
+				tr.runtime),
+		),
+		teststep.NewHTTPCheck(
+			genericContainer.Log,
+			"Git Function simple check through gateway",
+			gitCfg.InClusterURL,
+			poll, tr.expectedResponse)), nil
+}
+
+func azureDevOpsAuthSecretData(tr testRepo) (map[string]string, error) {
+	data := map[string]string{}
+	for _, key := range tr.secretEnvKeys {
+		value, ok := os.LookupEnv(key)
+		if !ok {
+			return nil, errors.New(fmt.Sprintf("failed to lookup environment variable '%v'", key))
+		}
+		// I want these values to be assigned explicitly
+		if key == azureDevOpsPassEnvName {
+			data["password"] = value
+		}
+		if key == azureDevOpsUserEnvName {
+			data["username"] = value
+		}
+	}
+	return data, nil
+}
+
+func githubAuthSecretData(tr testRepo) (map[string]string, error) {
```

Comment:
Is github different that other providers?

I think that we should have two methods to create two types of auth: `BasicAuth` and `SSHAuth` or we can even remove them

Reply:
I added two providers:
- github using ssh key.
- azure devops using basic auth.

I am creating two different functions for the them because:
- content of the data is different based on the auth method.
- environment variables used to pass the credentials.

The idea was to have simple clear structure per provider to make it easy to add new providers later. I will wrap everything in a single function.